compiler/rustc_data_structures/src/sync.rs

```diff
@@ -26,6 +26,30 @@ use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
 pub use std::sync::atomic::Ordering;
 pub use std::sync::atomic::Ordering::SeqCst;
 
+// 0: uninitialized
+// 1: single-thread mode
+// 2: multiple-thread mode
+static PARALLEL: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(0);
+
+#[inline]
+pub fn is_parallel() -> bool {
+    match PARALLEL.load(Ordering::Relaxed) {
+        1 => false,
+        2 => true,
+        _ => panic!("unknown parallel mode"),
+    }
+}
+
+// Only set by the `-Z threads` compile option
+pub fn set_parallel(parallel: bool) {
+    match (PARALLEL.load(Ordering::Relaxed), parallel) {
```

Comment:
This needs to use compare and swap to avoid races.

Reply:
Done